common/src/primitives/merkle/tree/mod.rs

```diff
@@ -0,0 +1,291 @@
+        let res = Self { tree };
+        Ok(res)
+    }
+
+    pub fn root(&self) -> H256 {
+        *self.tree.last().expect("By design, at least one element must exist")
+    }
+
+    pub fn total_node_count(&self) -> TreeSize {
+        self.tree
+            .len()
+            .try_into()
+            .expect("(total_node_count) By design, tree_size is always > 0")
+    }
+
+    pub fn leaf_count(&self) -> NonZeroUsize {
+        let tree_size = TreeSize::from_value(self.tree.len())
+            .expect("(leaf_count) Tree size valid by construction");
+        tree_size.leaf_count()
+    }
+
+    pub fn level_count(&self) -> NonZeroUsize {
+        let tree_size = TreeSize::from_value(self.tree.len())
+            .expect("(level_count) Tree size valid by construction");
+        tree_size.level_count()
+    }
+
+    pub fn absolute_index_from_bottom(
+        tree_size: TreeSize,
+        level_from_bottom: usize,
+        index_in_level: usize,
+    ) -> Option<usize> {
+        let node = NodePosition::from_position(tree_size, level_from_bottom, index_in_level)?;
+
+        Some(node.abs_index())
+    }
+
+    pub fn node_value_from_bottom(
+        &self,
+        level_from_bottom: usize,
+        index_in_level: usize,
+    ) -> Option<H256> {
+        let index_in_tree = Self::absolute_index_from_bottom(
+            self.tree.len().try_into().expect("Tree size is by design > 0"),
+            level_from_bottom,
+            index_in_level,
+        )?;
+
+        Some(self.tree[index_in_tree])
+    }
+
+    pub fn node_from_bottom(
+        &self,
+        level_from_bottom: usize,
+        index_in_level: usize,
+    ) -> Option<Node> {
+        let absolute_index = Self::absolute_index_from_bottom(
+            self.tree.len().try_into().expect("Tree size is by design > 0"),
+            level_from_bottom,
+            index_in_level,
+        )?;
+
+        Some(Node {
+            tree_ref: self,
+            absolute_index,
+        })
+    }
+
+    /// Given an absolute index in the flattened tree, return the level and index at that level in the form (level, index_at_level)
+    pub fn position_from_index(tree_size: TreeSize, absolute_index: usize) -> Option<NodePosition> {
+        NodePosition::from_abs_index(tree_size, absolute_index)
+    }
+
+    pub fn iter_from_leaf_to_root(
+        &self,
+        leaf_index: usize,
+    ) -> Result<MerkleTreeNodeParentIterator, MerkleTreeAccessError> {
+        let leaf_count = self.leaf_count().get();
+
+        if leaf_index >= leaf_count {
+            return Err(MerkleTreeAccessError::IterStartIndexOutOfRange(
+                leaf_index, leaf_count,
+            ));
+        }
+
+        let res = MerkleTreeNodeParentIterator {
+            node: Some(Node {
+                tree_ref: self,
+                absolute_index: leaf_index,
+            }),
+        };
+
+        Ok(res)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct Node<'a> {
```

Comment:
Unforuntately, this isn't possible, because both have different sources of information. Unless I create an enum of either a reference or tree size or some complex traits, this won't be possible.

Reply:
Unresolved this comment.
At the expense of doing many unnecessary reallocations, I did this... I don't like this, but I guess either this or reimpls.